### ISSUE TITLE
Add ability to use GitHub Actions to build NVDA

### ISSUE THREAD

**Issue by dpy013** [nvaccess/nvda#10516]
hi
The appveyor currently used by NVDA cannot compile multithreaded when compiling NVDA
### Is your feature request related to a problem? Please describe.
If possible, we can compile NVDA with GitHub Actions
### Describe the solution you'd like
help docs
[GitHub Actions docs](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions)
and
[Events that trigger workflows - GitHub](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/events-that-trigger-workflows)
### Describe alternatives you've considered
none
### Additional context

| Image | YAML Label | Included Software | Rollout Status of Latest Image Release |
| --------------------|---------------------|--------------------|--------------------|
| Ubuntu 24.04 <sup>beta</sup> | `ubuntu-24.04` | [ubuntu-24.04] | ![Endpoint Badge](https://img.shields.io/endpoint?url=https%3A%2F%2Fgist.githubusercontent.com%2Fijunaidm%2F0b5459bbe18604a639e1d2f7accd204f%2Fraw%2Fubuntu24.json) |
| Ubuntu 22.04 | `ubuntu-latest` or `ubuntu-22.04` | [ubuntu-22.04] | ![Endpoint Badge](https://img.shields.io/endpoint?url=https%3A%2F%2Fgist.githubusercontent.com%2Fijunaidm%2F0b5459bbe18604a639e1d2f7accd204f%2Fraw%2Fubuntu22.json) |
| Ubuntu 20.04 | `ubuntu-20.04` | [ubuntu-20.04] | ![Endpoint Badge](https://img.shields.io/endpoint?url=https%3A%2F%2Fgist.githubusercontent.com%2Fijunaidm%2F0b5459bbe18604a639e1d2f7accd204f%2Fraw%2Fubuntu20.json) |
| macOS 14 | `macos-latest-large` or `macos-14-large`| [macOS-14] | ![Endpoint Badge](https://img.shields.io/endpoint?url=https%3A%2F%2Fgist.githubusercontent.com%2Fijunaidm%2F0b5459bbe18604a639e1d2f7accd204f%2Fraw%2Fmacos-14.json) |
| macOS 14 Arm64 |`macos-latest`, `macos-14`, `macos-latest-xlarge` or `macos-14-xlarge`| [macOS-14-arm64] | ![Endpoint Badge](https://img.shields.io/endpoint?url=https%3A%2F%2Fgist.githubusercontent.com%2Fijunaidm%2F0b5459bbe18604a639e1d2f7accd204f%2Fraw%2Fmacos-14-arm64.json) |
| macOS 13 | `macos-13` or `macos-13-large` | [macOS-13] | ![Endpoint Badge](https://img.shields.io/endpoint?url=https%3A%2F%2Fgist.githubusercontent.com%2Fijunaidm%2F0b5459bbe18604a639e1d2f7accd204f%2Fraw%2Fmacos-13.json) |
| macOS 13 Arm64 | `macos-13-xlarge` | [macOS-13-arm64] | ![Endpoint Badge](https://img.shields.io/endpoint?url=https%3A%2F%2Fgist.githubusercontent.com%2Fijunaidm%2F0b5459bbe18604a639e1d2f7accd204f%2Fraw%2Fmacos-13-arm64.json) |
| macOS 12 | `macos-12` or `macos-12-large`| [macOS-12] | ![Endpoint Badge](https://img.shields.io/endpoint?url=https%3A%2F%2Fgist.githubusercontent.com%2Fijunaidm%2F0b5459bbe18604a639e1d2f7accd204f%2Fraw%2Fmacos-12.json) |
| Windows Server 2022 | `windows-latest` or `windows-2022` | [windows-2022] | ![Endpoint Badge](https://img.shields.io/endpoint?url=https%3A%2F%2Fgist.githubusercontent.com%2Fijunaidm%2F0b5459bbe18604a639e1d2f7accd204f%2Fraw%2Fwin22.json) |
| Windows Server 2019 | `windows-2019` | [windows-2019] | ![Endpoint Badge](https://img.shields.io/endpoint?url=https%3A%2F%2Fgist.githubusercontent.com%2Fijunaidm%2F0b5459bbe18604a639e1d2f7accd204f%2Fraw%2Fwin19.json) |
thanks

**Comment by dingpengyu** [comment]
cc @josephsl 
and
@michaelDCurran 
system config （E5 2vCPU/​7G RAM）

**Comment by leonardder** [comment]
I have investigated github actions for add-on compilation, and it wasn't suitable at that time because of https://github.com/actions/upload-artifact/issues/3. In short, Github actions uploads artifacts in a zip file. Even when that would be supported, wildcards aren't either, see https://github.com/actions/upload-artifact/issues/11.

Having said that, it will probably take hours or days to convert our appveyor script to Github Actions, thereby not gaining much benefit. Closing as can't fix for now. We can always reopen if desired.

**Comment by dingpengyu** [comment]
> I have investigated github actions for add-on compilation, and it wasn't suitable at that time because of [actions/upload-artifact#3](https://github.com/actions/upload-artifact/issues/3). In short, Github actions uploads artifacts in a zip file. Even when that would be supported, wildcards aren't either, see [actions/upload-artifact#11](https://github.com/actions/upload-artifact/issues/11).
> 
> Having said that, it will probably take hours or days to convert our appveyor script to Github Actions, thereby not gaining much benefit. Closing as can't fix for now. We can always reopen if desired.
Thank you for your instructions

**Comment by XLTechie** [comment]
One of my back burner projects has been to convert the Appveyor script to 
something more generic.
The goal being so that people who fork the main repo, do not actually need a 
source compilation environment to test their own PRs and private changes, or 
otherwise build from source for their own purposes.

I haven't gotten far with it because of lack of time, but it seems to me that if 
we could get this working as an Appveyor alternative, it would make that idea 
ultimately a bit more straight forward.

**Comment by leonardder** [comment]
I agree that appveyor is not very ideal, however, our current appveyor 
script uses pretty specific functionality, also related to tests. It 
will take some time before actions is mature enough.

**Comment by dingpengyu** [comment]
cc @seanbudd

**Comment by dingpengyu** [comment]
> I have investigated github actions for add-on compilation, and it wasn't suitable at that time because of [actions/upload-artifact#3](https://github.com/actions/upload-artifact/issues/3). In short, Github actions uploads artifacts in a zip file. Even when that would be supported, wildcards aren't either, see [actions/upload-artifact#11](https://github.com/actions/upload-artifact/issues/11).
> 
> Having said that, it will probably take hours or days to convert our appveyor script to Github Actions, thereby not gaining much benefit. Closing as can't fix for now. We can always reopen if desired.

Has the above problem been solved by github actions?

**Comment by xingkong0113** [comment]
appveyor has started to limit the number of downloads to.
{"message":"Artifacts download limit (1024 MB/day) exceeded."}
test method is as follows: 
first open this [address](https://ci.appveyor.com/api/buildjobs/tr6nwjw1ucmg6far/artifacts/output%2Fnvda_snapshot_alpha-24523%2C7735d962.exe)
Once opened, the above prompt will appear
{"message":"Artifacts download limit (1024 MB/day) exceeded."}

**Comment by leonardder** [comment]
See #13281

**Comment by leonardder** [comment]
As actions is much more mature now and a switch to actions could be a possible way to avoid #13281, I"ll reopen this until a solution for #13281 has been chosen.

**Comment by leonardder** [comment]
never mind, this is still a showstopper for us:

> During a workflow run, files are uploaded and downloaded individually using the upload-artifact and download-artifact actions. However, when a workflow run finishes and an artifact is downloaded from either the UI or through the download api, a zip is dynamically created with all the file contents that were uploaded. There is currently no way to download artifacts after a workflow run finishes in a format other than a zip or to download artifact contents individually.

**Comment by dpy013** [comment]
> I have investigated github actions for add-on compilation, and it wasn't suitable at that time because of [actions/upload-artifact#3](https://github.com/actions/upload-artifact/issues/3). In short, Github actions uploads artifacts in a zip file. Even when that would be supported, wildcards aren't either, see [actions/upload-artifact#11](https://github.com/actions/upload-artifact/issues/11).
> 
> Having said that, it will probably take hours or days to convert our appveyor script to Github Actions, thereby not gaining much benefit. Closing as can't fix for now. We can always reopen if desired.
hi Maybe [Creepios/sftp-action: GitHub Action to upload files/directories via ssh](https://github.com/Creepios/sftp-action) This problem can be solved

**Comment by XLTechie** [comment]
@LeonarddeR Your show-stopper seems no longer to be the case. Can you confirm that your concerns are satisfied?

**Comment by LeonarddeR** [comment]
After https://github.com/nvaccess/nvda/issues/13281, snapshots are no longer directly downloaded from appveyor. This means that the fact that github only offers zip files is less relevant, though it is still a bit of a complicating factor when downloading artifacts from pull requests. That said, I think there's enough potential to reopen this.

**Comment by seanbudd** [comment]
Hi, just to give an update for NV Access. Moving to GitHub actions is on our roadmap due to many issues with AppVeyor:
- AppVeyor instances appear to be getting slower, slowing the build times
- We can only perform one build at a time
- GitHub actions allows us to split up the CI/CD pipeline (e.g. the lint check can run independently of the build)
- GitHub actions are easier for forks to use
- being in one environment (i.e. GitHub) is generally simpler

**Comment by XLTechie** [comment]
@seanbudd wrote:

>  *  AppVeyor instances appear to be getting slower, slowing the build times

We have long had the ability to make those significantly faster, by caching already compiled components and deploying them to the run environment through AppVeyor's caching options.

I often wondered why NV Access didn't, especially for the slow-to-compile C components, and clonable items that hadn't been updated, but assumed there was a reason.

Of course, Actions can do that, too.

**Comment by dpy013** [comment]
As a side note, GitHub actions can be used to compile and test NVDA code using relatively new windows systems.
It would also be possible to link crowdin to the GitHub code repository.
Finally, it might be possible to automate the task of posting alpha, stable, and test boards to the website.
To reduce the workload of the staff!

**Comment by seanbudd** [comment]
@XLTechie 

>I often wondered why NV Access didn't, especially for the slow-to-compile C components, and clonable items that hadn't been updated, but assumed there was a reason.

We've already done this in a different way for some components like https://github.com/nvaccess/nvda-cldr 

The issue of AppVeyor making their instances cheaper and slower over time is still going to be an issue even if we optimize other parts of the build system. The slowest parts of the build are quite challenging to split up. We can always speed up the build system in many ways, and GitHub actions is probably the most effective solution.

**Comment by XLTechie** [comment]
@dpy013 You are right about the relatively new Windows systems. It might even allow for broader testing setups, if we can cause artifacts to be tested on several versions of Windows at once.  That alone would probably make it worth doing.

Certainly, if Crowdin can be linked to AppVeyor compiles as it is now, it can be equally well linked to Actions compilations, probably with the same scripts.

Posting artifacts to the website is already automated, at least for alphas and betas.

**Comment by nvdaes** [comment]
I've started to study modules of GitHub Actions with the Study guide for a possible preparation to get the GitHub Actions certificate from GitHub. I don't know if finally I'll try to get the certifications since I need to show Government documentation, installing a software etc. But if someone doesn't take this, I may create a PRafter testing.
We may use the new attestations feature too.

**Comment by seanbudd** [comment]
@nvdaes - there's no need to worry about build signing, we're currently working on changing this actively

**Comment by dpy013** [comment]
I happened to have nothing to do in the morning, so I went and converted a version of GitHub actions using gpt4 + NVDA's appveyor.yml.
Now stuck at appveyor\scripts\setBuildVersionVars.ps1
I'm stuck on this script, plus I don't know much about PowerShell Core, so anyone interested can take it and look into it.
https://github.com/dpy013/nvda/blob/1c8e80ff50c4371dae40908ddcfcd45b9f0283c5/.github/workflows/NVDABuild.yaml

**Comment by dpy013** [comment]
Currently, the one-hour time limit on AppVeyor is insufficient for compiling NVDA alpha-35333. Does the official team have any plans to implement issue 10516?
If this issue is not needed, I'll prepare to convert this problem into a discussion.

**Comment by seanbudd** [comment]
@dpy013 this issue has been triaged - i.e. accepted. We plan to eventually move to GitHub actions. Community support to get this done is welcome. Please do not close or convert this issue.
Please also note that alpha builds fail sometimes, and the recent failure is not related to build time.

**Comment by nvdaes** [comment]
I think that I would work on this. I used GitHub Actions previously to test add-ons with NVDA and I didn't feel comfortable to use Appveyor.
Then we can discuss problems like build signing, and where to store artifacts.
Recently I studied GitHub foundations module, in addition to GitHub Actions studied previously, though I didn't take GitHub certifications, and this may be useful to practice.

**Comment by nvdaes** [comment]
Finally I have been able to use setup Python action to install Python 3.11 x64, using windows-2025 runner on GitHub. I'll try to use windows-latest again in case.
I will open a PR, though I'll require assistance to know which tags for system tests are required, at least.

**Comment by nvdaes** [comment]
I've been testing GitHub actions using powershell scripts used for appveyor, but seems that this doesn't work. Also, I've tested using the .bat files, like done in local computers, and this works.
Some Chrome system tests are always failing, and I've seen some staff in powershell so that appveyor considers this fact.
After testing, my approach for a preliminary PR would be to use the GitHub script action instead of Powershell to set the appropriate version, if it's a release etc., depppending on the GitHub context (sha, branch name, etc). And I'll use the .bat files, not powershell.
For things like webhooks, signing or upload symbols to Mozilla, I will stop since this needs help from NV Access.
I comment this in case you prefer another approach. Appveyor is used years ago and I had thought about making a migration with equivalent commands, but seems that Powershell scripts are difficult to use.

**Comment by nvdaes** [comment]
Finally I'm using powershell scripts when possible. A PR is opened (work in progress for now).

**Comment by dpy013** [comment]
> I have investigated github actions for add-on compilation, and it wasn't suitable at that time because of [actions/upload-artifact#3](https://github.com/actions/upload-artifact/issues/3). In short, Github actions uploads artifacts in a zip file. Even when that would be supported, wildcards aren't either, see [actions/upload-artifact#11](https://github.com/actions/upload-artifact/issues/11).
          > 
          > Having said that, it will probably take hours or days to convert our appveyor script to Github Actions, thereby not gaining much benefit. Closing as can't fix for now. We can always reopen if desired.
          
          According to the latest news, the main branch of upload-artifact GitHub actions already supports direct file uploads, but we may have to wait until version 6.2.0 to use this feature. For details, please refer to:
https://github.com/actions/upload-artifact/pull/764

**Comment by XLTechie** [comment]
I have been using GitHub actions to upload nvda-addon files for years.

See, for example

https://github.com/opensourcesys/numpadnavmode/blob/master/.github/workflows/build_addon.yml

On push, it builds the add-on, and uploads a release containing the .nvda-addon file, a .pot file, and the usual release artifacts.
